lib/datafile.js

```diff
@@ -148,21 +148,29 @@ _.extend(Block.prototype, {
     },
 
     _canReUseBuffer: function(size) {
-        return this.reUseBuffer && this._readOffset >= size;
+        return this.reUseBuffer && (this._buffer.length - this.remainingBytes) >= size;
     },
 
     _resizeIfRequired: function(size) {
         if (this._canReUseBuffer(size)) {
-            if (this._readOffset != this._writeOffset)
+            if (this._readOffset > 0 && this._readOffset != this._writeOffset) {
                 this._buffer.copy(this._buffer, 0, this._readOffset, this._writeOffset);
-            this._writeOffset = this.remainingBytes;
-            this._readOffset = 0;
-        } else if (this._writeOffset + size > this._buffer.length) {
+            }
+        } else {
             var oldBuffer = this._buffer;
-            this._buffer = new Buffer(this._bufferSize(size));
-            oldBuffer.copy(this._buffer, 0);
+
+            if (this.remainingBytes + size > this._buffer.length) {
```

Comment:
Can this be instead `if (this.reUseBuffer)`? Although it probably isn't semantically correct, which would be more important than the (minuscule) optimization we would gain by the change...